File: LeetCode/130SurroundedRegions/SurroundedRegions.cpp
class Solution {
public:
    void dfs(vector<vector<char>>& board, int r, int c) {
        if (r < 0 || c < 0 || r >= board.size() || c >= board[0].size() || board[r][c] != 'O') {
            return;
        }
        board[r][c] = 'T';
        dfs(board, r-1, c);
        dfs(board, r+1, c);
        dfs(board, r, c-1);
        dfs(board, r, c+1);
    }
    void solve(vector<vector<char>>& board) {
        // DFS -> Time:O(n), Space:O(n^2)

        int rows = board.size(), cols = board[0].size();

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (i == 0 || i == rows-1 || j == 0 || j == cols-1) {
                    dfs(board, i, j);
                }
            }
        }

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
                if (board[i][j] == 'T') {
                    board[i][j] = 'O';
                }
            }
        }
        
    }
};

File: LeetCode/130SurroundedRegions/SurroundedRegions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        # DFS -> Time:O(n), Space:O(n^2)

        ROWS, COLS = len(board), len(board[0])

        def dfs(board,r,c):
            if (r < 0 or c < 0 or r >= ROWS or c >= COLS or board[r][c] != "O"):
                return
            board[r][c] = 'T'
            dfs(board, r-1, c)
            dfs(board, r+1, c)
            dfs(board, r, c-1)
            dfs(board, r, c+1)

        for i in range(ROWS):
            for j in range(COLS):
                if ((i == 0 or i == ROWS-1) or (j == 0 or j == COLS-1)) and board[i][j] == "O":
                    dfs(board, i, j)

        for i in range(ROWS):
            for j in range(COLS):
                if board[i][j] == "O":
                    board[i][j] = "X"
                if board[i][j] == "T":
                    board[i][j] = "O"
